src/main/java/com/farion/onlinebookstore/validator/CategoryIdsValidator.java

```diff
@@ -0,0 +1,30 @@
+package com.farion.onlinebookstore.validator;
+
+import com.farion.onlinebookstore.lib.CategoryIds;
+import com.farion.onlinebookstore.service.CategoryService;
+import com.farion.onlinebookstore.util.ConditionHolder;
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+import java.util.Set;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Component;
+
+@RequiredArgsConstructor
+@Component
+public class CategoryIdsValidator implements ConstraintValidator<CategoryIds, Set<Long>> {
+    private final ConditionHolder conditionHolder;
+    private final CategoryService categoryService;
+
+    @Override
+    public void initialize(CategoryIds constraintAnnotation) {
+        conditionHolder.setCondition(ThreadLocal.withInitial(categoryService::getAllCategoryIds));
+    }
+
+    @Override
+    public boolean isValid(
+            Set<Long> categoryIds,
+            ConstraintValidatorContext constraintValidatorContext
+    ) {
+        return conditionHolder.getCondition().get().containsAll(categoryIds);
+    }
+}
```

Comment:
Annotation-validations in DTO are usually lightweight and do not have any service or repository involved checks, but some more simple ones, like pattern matching, or null/not null etc. because imagine if you have 100 fields like that, you'd be making 100 sessions/transactions/db calls to check all that individually which is not very efficient
Such checks on existence should be done in services, and that validator should be removed along with the `ConditionHolder`